### PR TITLE
Send kafka message when refund is created

### PR DESCRIPTION
**Pull request by mateuszc-kainos**
Send kafka message when refund is created

https://companieshouse.atlassian.net/browse/S4-254

### Type of change

* [ ] Bug fix
* [x] New feature
* [ ] Breaking change

### Pull request checklist

* [x] I have added unit tests for new code that I have added
* [x] I have added/updated functional tests where appropriate
* [x] The code follows our [coding standards](https://github.com/companieshouse/styleguides/blob/master/go.md)

__An exhaustive list of peer review checks can be read [here](https://github.com/golang/go/wiki/CodeReviewComments)__

### CODE REVIEW

File: handlers/callback.go [hunk 1]
```diff
@@ -12,7 +12,7 @@ import (
 )
 
 // handleKafkaMessage allows us to mock the call to produceKafkaMessage for unit tests
-var handleKafkaMessage = produceKafkaMessage
+var handlePaymentMessage = ProducePaymentMessage
```
Comment (peadarkelly) [inline handlers/callback.go:15]: why capital P in ProducePaymentMessage ?

File: handlers/refunds.go [hunk 2]
```diff
@@ -63,4 +65,11 @@ func HandleCreateRefund(w http.ResponseWriter, req *http.Request) {
 	}
 
 	log.InfoR(req, "Successful POST request for new refund", log.Data{"refund_id": refund.RefundId, "status": http.StatusCreated})
+
+	err = handleRefundMessage(paymentResource.MetaData.ID, refund.RefundId)
+	if err != nil {
+		log.ErrorR(req, fmt.Errorf("error producing kafka message: [%v]", err))
```
Comment (peadarkelly) [inline handlers/refunds.go:71]: might be good to differentiate between payment kafka message errors vs refund kafka message errors.

File: service/refund_test.go [hunk 3]
```diff
@@ -38,7 +38,7 @@ func TestUnitCreateRefund(t *testing.T) {
 			GetGovPayRefundSummary(req, id).
 			Return(nil, nil, Error, err)
 
-		refund, status, err := service.CreateRefund(req, id, body)
+		_, refund, status, err := service.CreateRefund(req, id, body)
```
Comment (peadarkelly) [inline service/refund_test.go:41]: if paymentSession is now being returned from the service, the unit tests should assert this value at least once.

File: handlers/callback.go [hunk 4]
```diff
@@ -12,7 +12,7 @@ import (
 )
 
 // handleKafkaMessage allows us to mock the call to produceKafkaMessage for unit tests
```
Comment (BSaunders95) [inline handlers/callback.go:14]: This comment is no longer applicable given a change in the variable name

File: handlers/kafka_helper.go [hunk 5]
```diff
@@ -22,6 +22,15 @@ const ProducerSchemaName = "payment-processed"
 // paymentProcessed represents the avro schema which can be found in the chs-kafka-schemas repo
 type paymentProcessed struct {
 	PaymentSessionID string `avro:"payment_resource_id"`
+	RefundId         string `avro:"refund_id,omitempty"`
+}
+
+func producePaymentMessage(paymentID string) error {
+	return produceKafkaMessage(paymentID, "")
```
Comment (peadarkelly) [inline handlers/kafka_helper.go:29]: I thought we didn;t have to add an empty string as the refundId on the Payment message? I thought we could ignore it from the Payments API no bother, but it was get serialised to an empty string when consuming the kafka message?

File: service/refund_test.go [hunk 6]
```diff
@@ -38,8 +38,9 @@ func TestUnitCreateRefund(t *testing.T) {
 			GetGovPayRefundSummary(req, id).
 			Return(nil, nil, Error, err)
 
-		refund, status, err := service.CreateRefund(req, id, body)
+		paymentSession, refund, status, err := service.CreateRefund(req, id, body)
 
+		So(paymentSession, ShouldBeNil)
```
Comment (peadarkelly) [inline service/refund_test.go:43]: why is paymentSession always Nil? looking at these tests I would assume that it's always Nil regardless of the test. Please vary your assertions in unit tests for maximum coverage of different scenarios.

File: handlers/kafka_helper.go (hunk 5 above)
Comment (mateuszc-kainos) [reply handlers/kafka_helper.go:29]: Would you see it as a separate method then? There are no optional variables in Go.

File: service/refund_test.go (hunk 6 above)
Comment (mateuszc-kainos) [reply service/refund_test.go:43]: Whenever errors occur, paymentSession is nil. In the last, positive test the paymentSession is not nil.

File: handlers/kafka_helper.go (hunk 5 above)
Comment (peadarkelly) [reply handlers/kafka_helper.go:29]: ah if it's going to require a new method then forget about it.

**Review by peadarkelly — APPROVED** [review]

**Pull request merged**